docs/tutorials/paroutes_benchmark.ipynb

```diff
@@ -0,0 +1,761 @@
+    "inventory = PaRoutesInventory(n=PAROUTES_N)\n",
+    "print(f\"Size of inventory: {len(inventory.purchasable_mols())}\")\n",
+    "print(\"First few molecules in inventory:\")\n",
+    "pprint(list(inventory.purchasable_mols())[:5])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "id": "256137ca-e572-4840-a47f-9ab386799c5a",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Number of test SMILES: 10000\n",
+      "First test SMILES:\n",
+      "['COc1ccc(F)c(-c2ccc(COc3cc(C(CC(=O)O)C4CC4)ccc3F)nc2CC(C)(C)C)c1',\n",
+      " 'CCc1cc2nncc(N3CCc4[nH]nc(C(=O)NC5CC5)c4C3)c2cc1OC',\n",
+      " 'Clc1ccc(-n2cncn2)c(Cc2nc3c(NCCC4CCCCN4)nccc3o2)c1',\n",
+      " 'CCOCc1nc2c(N)nc3ccc(O)cc3c2n1CC(C)(C)NS(C)(=O)=O',\n",
+      " 'O=C(O)CCNC(=O)c1nc(-c2ccncc2)c2c(cc(-c3ccccc3)c(=O)n2CC2CCCCC2)c1O']\n"
+     ]
+    }
+   ],
+   "source": [
+    "# Load test smiles\n",
+    "from syntheseus_paroutes import get_target_smiles_list\n",
+    "test_smiles = get_target_smiles_list(n=PAROUTES_N)\n",
+    "print(f\"Number of test SMILES: {len(test_smiles)}\")\n",
+    "print(\"First test SMILES:\")\n",
+    "pprint(test_smiles[:5])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "3359f01d-ddd1-4ad9-857e-9b93c787c074",
+   "metadata": {},
+   "source": [
+    "## Step 2: set up algorithms"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "id": "f25c0a2a-9bbc-4692-a384-93eb219d3261",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Some constants for all algorithms\n",
+    "RXN_MODEL_CALL_LIMIT = 100\n",
+    "TIME_LIMIT_S = 300"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "df4f6506-0dcd-4296-b590-994ea69e9fa3",
+   "metadata": {},
+   "source": [
+    "### 2a: Retro-star\n",
+    "\n",
+    "Retro-star is a best-first search algorithm for AND/OR trees\n",
+    "which tries to find minimum-cost routes.\n",
+    "It requires 3 cost functions to be provided to the algorithm.\n",
+    "\n",
+    "1. A function to give the cost of each molecule (OrNode)\n",
+    "2. A function to give the cost of each reaction (AndNode)\n",
+    "3. A function to estimate the future cost to synthesize molecules, which is used as the search heuristic. This is called the \"reaction number\" in the original paper."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "id": "1309f29d-d935-40f1-b673-67dfa2cd9b21",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from syntheseus.search.algorithms.best_first import retro_star\n",
+    "from syntheseus.search.graph.and_or import AndNode, OrNode, AndOrGraph\n",
+    "from syntheseus.search.node_evaluation.common import ReactionModelLogProbCost"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "id": "c891cc66-016e-456b-ada5-5f9f980e07a2",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#1: OrNode cost function.\n",
```

Comment:
nit: maybe add a space after "#"? (if yes then there are five other places like this)